File: scripts/sim_edu_bbt/funcs/modulator.m
function [x_n, mis] = modulator(byte_seq, spar)
  % MODULATOR Converts a byte sequence into a modulated signal.
  %
  % [x_n, mis] = modulator(byte_seq, spar) takes a byte sequence and
  % converts it into a modulated signal by applying upsampling and pulse shaping.
  %
  % Inputs:
  %   byte_seq - Byte sequence to be modulated (values between 0 and 255)
  %   spar     - Structure of parameters with:
  %              * n_pre  : Number of bits in the preamble
  %              * n_sfd  : Number of bits in the Start Frame Delimiter (SFD)
  %              * n_pulse: Upsampling factor
  %              * pulse  : Pulse shaping waveform
  %
  % Outputs:
  %   x_n - Final modulated signal
  %   mis - Structure with internal signals used in modulation
  
    % Verify that all values in byte_seq are valid bytes (0-255)
    assert(all(byte_seq <= 255));
  
    n_bytes = length(byte_seq);
  
    % Generate preamble (alternating 0s and 1s)
    pre          = zeros(1, spar.n_pre);
    pre(2:2:end) = 1;
  
    % Generate Start Frame Delimiter (SFD)
    sfd = zeros(1, spar.n_sfd);
    if mod(spar.n_pre, 2) == 0
      sfd(1:2:end) = 1;
    else
      sfd(2:2:end) = 1;
    end
  
    % Convert bytes to bits
    data = [];
    for i = 1:n_bytes
      bin_str = dec2bin(byte_seq(i), 8); % Convert to binary (8 bits)
      bin = str2num(bin_str(:))';        % Convert string to numeric vector
      data = [data bin];
    end
  
    % Construct packet with preamble, SFD, and data
    packet = [pre sfd data];
  
    % Map bits to symbols (-1, 1)
    x = 2 * packet - 1;
  
    % Pulse shaping
    xx  = upsample(x, spar.n_pulse); % Upsampling with factor spar.n_pulse
    xxx = conv(xx, spar.pulse);      % Filtering to smooth the signal
  
    % Function outputs
    x_n = xxx;  % Modulated signal
    mis.d = packet; % Internal signal of original bits
  
  end
  

File: ejercicios/TpFinal/lib/modulator/modulator.py
## @file modulator.py
# @brief Module containing the Modulator class for digital signal modulation.
# @details This module implements a Modulator class that handles packet creation and modulation
# for a digital communication system. It uses different pulse shapes for modulation and includes
# functionality for adding preamble and start frame delimiter (SFD) sequences.
# @author Marcos Dominguez
# @author Lucas Monzon Languasco

import numpy as np
from lib.pulse_generator.pulse_generator import PulseGenerator
## @class Modulator
#  @brief Handles the modulation process for a byte sequence.
#  @details Uses a specified pulse shape to modulate binary sequences derived from byte arrays.
#           Adds preamble and start frame delimiter (SFD) sequences before the data bits.
#           Produces a baseband modulated signal using pulse shaping.
class Modulator:
    """
    @brief Handles the modulation process for a byte sequence.
    """

    ## @brief Constructor for Modulator
    #  @param n_bytes Number of bytes in the payload
    #  @param n_pre Number of bits in the preamble
    #  @param n_sfd Number of bits in the start frame delimiter (SFD)
    #  @param pulse Name of the pulse shape to use (passed to PulseGenerator)
    #  @param Ts Sampling interval
    #  @param Tsymb Symbol duration
    def __init__(self, n_bytes, n_pre, n_sfd, pulse:str, Ts, Tsymb):
        self._n_bytes = n_bytes
        self._n_pre = n_pre
        self._n_sfd = n_sfd
        self._pulse = PulseGenerator(pulse, Ts, Tsymb)

    ## @brief Creates a packet with preamble, SFD, and data bits
    #  @param byte_seq Sequence of bytes to encode
    #  @return A tuple (x, data) where x is the bipolar sequence and data is the binary array of payload bits
    def create_packet(self, byte_seq):
        pre = np.zeros(self._n_pre, dtype=int)
        pre[1::2] = 1
        sfd = np.zeros(self._n_sfd, dtype=int)
        if self._n_pre % 2 == 0:
            sfd[0::2] = 1
        else:
            sfd[1::2] = 1
        data = np.hstack([np.array(list(np.binary_repr(byte, 8)), dtype=int) for byte in byte_seq])
        packet = np.hstack([pre, sfd, data])
        x = 2 * packet - 1
        return x, data

    ## @brief Modulates a sequence of bytes into a baseband waveform
    #  @param byte_seq Sequence of bytes to modulate (values must be in [0, 255])
    #  @return The modulated signal (convolved with pulse shape)
    def modulate(self, byte_seq):
        assert np.all(np.array(byte_seq) <= 255), "All values in byte_seq must be between 0 and 255"
        x, self._bits = self.create_packet(byte_seq)
        xx = np.zeros(len(x) * self._pulse.n_pulse)
        xx[::self._pulse.n_pulse] = x
        xxx = np.convolve(xx, self._pulse.pulse)
        pad_size = (len(xxx) - len(xx)) // 2
        self._d = np.pad(xx, (pad_size, len(xxx) - len(xx) - pad_size), mode='constant')
        return xxx

    ## @brief Returns number of bytes in the payload
    @property
    def n_bytes(self):
        return self._n_bytes

    ## @brief Returns number of bits in the preamble
    @property
    def n_pre(self):
        return self._n_pre

    ## @brief Returns number of bits in the start frame delimiter (SFD)
    @property
    def n_sfd(self):
        return self._n_sfd

    ## @brief Returns number of samples per symbol (pulse length)
    @property
    def n_pulse(self):
        return self._pulse.n_pulse

    ## @brief Returns the pulse shape used for modulation
    @property
    def pulse(self):
        return self._pulse.pulse

    ## @brief Returns the sampling interval
    @property
    def Ts(self):
        return self._pulse.Ts

    ## @brief Returns the symbol duration
    @property
    def Tsymb(self):
        return self._pulse.Tsymb

    ## @brief Returns the length of the FIR pulse filter
    @property
    def n_fir(self):
        return self._pulse._n_fir

    ## @brief Returns the binary payload bits
    @property
    def bits(self):
        return self._bits

    ## @brief Returns the zero-padded version of the bipolar signal before convolution
    @property
    def d(self):
        return self._d
